Fix-72650 Shell paths should be validated before launched to prevent ambiguous errors - $Path check

skprabhanjan opened the pull request:
@Tyriar , Added a trivial path check to fix #72650  , Please review if this is good or we need to handle it differently and in a more better way than this naive approach, Thanks :)

Tyriar commented on src/vs/workbench/contrib/terminal/node/terminalProcess.ts:
```diff
@@ -69,16 +69,17 @@ export class TerminalProcess implements ITerminalChildProcess, IDisposable {
+		let envPathVariable: string | undefined = process.env.PATH;
+
+		fs.stat(shellLaunchConfig.executable!, (err) => {
+			if (err && err.code === 'ENOENT' && envPathVariable!.search(shellLaunchConfig.executable!) === -1) {
```
This won't work because executable is a file and $PATH just contains a bunch of folders, it would need to check the contents of every folder inside $PATH for executable, iff it's not an absolute path.

skprabhanjan replied:
Done.

skprabhanjan commented:
@Tyriar , I had to achieve this with quite some added logic with the use of callbacks because the nature of the functions we needed were all async.
Please review these new changes and let me know if it is good or we need to improve it more, Thanks :)

skprabhanjan commented:
@Tyriar.
**UPDATE** : I removed the callback logic as it was getting messy and complicated with async funcs.
Changed it to be sync , Please check if this is good and review the code , Thanks :)

skprabhanjan commented:
@Tyriar , Did you get a chance to check this and review ?

Tyriar commented:
The general structure looks good, will do a proper review/merge soon (little busy atm).

skprabhanjan commented:
@Tyriar , Did you get a chance to review this PR ?

Tyriar commented:
I'll make a note to check it out on tomorrow when master should be open again, hopefully we can get it in for the July release 😃

Tyriar commented on src/vs/workbench/contrib/terminal/node/terminalProcess.ts:
```diff
@@ -69,16 +70,48 @@ export class TerminalProcess implements ITerminalChildProcess, IDisposable {
+
+	private checkInsideDirRecursively(execPath: string, executable: string): void {
+		try {
+			fs.readdirSync(execPath).forEach((file) => {
```
We can't use any sync calls

Tyriar replied:
We have `fs` call in a `pfs` file which "promisifies" them, that's the preferred way to do async file access.

Tyriar commented on src/vs/workbench/contrib/terminal/node/terminalProcess.ts:
```diff
@@ -69,16 +70,48 @@ export class TerminalProcess implements ITerminalChildProcess, IDisposable {
+		try {
+			fs.readdirSync(execPath).forEach((file) => {
+				const pathToFile = path.join(execPath, file);
+				const isDirectory = fs.statSync(pathToFile).isDirectory();
```
Another sync call

Tyriar commented:
Oh you unchecked letting maintainers push to your branch?

Tyriar commented on src/vs/workbench/contrib/terminal/node/terminalProcess.ts:
```diff
@@ -69,16 +70,48 @@ export class TerminalProcess implements ITerminalChildProcess, IDisposable {
-		this.setupPtyProcess(shellLaunchConfig, options);
-		// });
+		fs.stat(shellLaunchConfig.executable!, (err) => {
+			if (err && err.code === 'ENOENT' && !this.checkIfExistsInPath(shellLaunchConfig.executable!)) {
```
This doesn't seem to work either:

```
"terminal.integrated.shell.osx": "bash",
```

<img width="452" alt="Screen Shot 2019-06-28 at 11 26 20 AM" src="https://user-images.githubusercontent.com/2193314/60363159-94f97a80-9997-11e9-9e28-f3fca0bd56ba.png">

skprabhanjan commented:
> Oh you unchecked letting maintainers push to your branch?

Nope , it is still checked

skprabhanjan commented:
@Tyriar , So would you suggest me to do currently ?

Tyriar commented:
@skprabhanjan I was unable to push to your branch, could you sync the branch with master? Maybe I was trying to push too many commits or something 😕 

I can try finish things off if I can push again

skprabhanjan commented:
Seems like the branch was outdated but it was showing already up to date .
Created a new branch and closing this PR and sending a new one.
Thanks for the review :)
